### DISCUSSION TITLE
Relation between Controller, Actor, Strategy

### DISCUSSION THREAD

**Discussion by HeidXC** [nautechsystems/nautilus_trader#1627] [Q&A]
Dear maintainers,
When I read about controller, actor and strategy they seem have very similar functionality. I can imagine controller to take the responsibility of starting/ stopping strategy based on  market states, what about actor? Is single strategy manage all data and execution details a preferred mode or is a controller manage multiple simple strategy and schedule them a preferred mode?

**Comment by fredmonroe** [comment]
I'm curious about this too. In particular the Controller class because I couldn't find a lot of examples of its use.

based on this [test ](https://github.com/nautechsystems/nautilus_trader/blob/380169ce19f6609d9d16404ad58338a4b69ccef5/nautilus_trader/test_kit/mocks/controller.py#L26)

it looks like its maybe a way to wrap trader objects with extra functionality, almost like a decorator function in python 
I think your idea about starting/stopping based on state is a good one - you could have "big red panic putton" that a controller thats attached to multiple traders executes (like get flat, or pull orders, etc)

**Comment by rsmb7z** [comment]
The Actor class provides a base class for writing any component or subclass you would like to integrate with the system. It has access to the Cache, Clock, and includes methods for data subscriptions and historical data requests. This is useful for any subclass that requires processing but not actual trading.

The Strategy class is a subclass of Actor and includes additional features for order execution.

The Controller class is also a subclass of Actor, but its purpose is to allow you to create, start, and stop other Actors and Strategies during runtime. This is useful when it may not be feasible to have strategies active for many stocks, which would require active data subscriptions for each strategy. When you need to deploy a strategy based on an event or external signal, such as from scanners, you can create and run the strategy for the required instruments as needed.

**Reply by fredmonroe** [reply]
thank you for this response its very timely for me, i'm working on something like this where i don't know until the day starts which symbols i'll need to execute on. 

related question - If I want to create custom shared data feeds - things like proprietary signals that fire off or maybe the output of something similar to a scanner - and where the data doesn't fit nicely into existing structures like bars or ticks, etc - how would you deploy that?

As an actor? And should i push the data out onto the message bus or stick it in the cache somehow?

**Reply by rsmb7z** [reply]
> related question - If I want to create custom shared data feeds - things like proprietary signals that fire off or maybe the output of something similar to a scanner - and where the data doesn't fit nicely into existing structures like bars or ticks, etc - how would you deploy that?
> 
> As an actor? And should i push the data out onto the message bus or stick it in the cache somehow?

To handle data feeds, it is more appropriate to write adapters using the LiveMarketDataClient and CustomData types, depending on your needs. These types will integrate seamlessly similar to other market data.

The Actor can subscribe to these data feeds and perform any necessary processing before generating further signals using `Actor.publish_signal`. You can add data to the cache using Cache.add(key, value).

While using an adapter is the recommended approach for handling data feeds, for more straightforward tasks such as scanning data from sources like Finviz, you can directly utilize the Actor. The Actor can process the DataFrame and generate signals efficiently. However, for any blocking operations, be sure to use an executor.
